### PR TITLE
Add support for Xiaomi Air purifier 3C 

### PR DESCRIPTION
**Pull request by arturdobo**
This PR adds support for Air purifier 3C (`zhimi.airpurifier.mb4`) which according to  [specification](https://miot-spec.org/miot-spec-v2/instance?type=urn:miot-spec-v2:device:air-purifier:0000A007:zhimi-mb4:2) provides only a subset of i.e. mb3 functionalities.

### CODE REVIEW

**Comment by pawelnguyen** [comment]
I bought a 3C and just found out it doesn't work with home assistant. LMK if I can help and test this on a real device.

Looking at the code - feature set seems correct, I can confirm that this model doesn't provide temperature/humidity.

Thanks for working on this!

File: miio/airpurifier_miot.py
Comment (rytilahti) [inline miio/airpurifier_miot.py:281]: Please add the model number presenting the device to the container docstring.

Also, add an example contents for the data structure, something like this: https://github.com/rytilahti/python-miio/blob/master/miio/airconditioner_miot.py#L178

File: miio/airpurifier_miot.py
Comment (rytilahti) [inline miio/airpurifier_miot.py:284]:
```suggestion
```
No need for an empty `__init__`.

**Review by rytilahti — CHANGES_REQUESTED** [review]
Looks straightforward to me, thanks for the PR!

A couple of minor issues should be changed and then this is ready to be merged.

**Comment by syssi** [comment]
Cool! Let's merge this PR as soon the minor issues are fixed.

**Comment by jagub2** [comment]
Howdy, I tried to use these changes, but I'm struggling to get it working.

```
 ~/python-miio/miio   add-mb4-support ± python3 cli.py airpurifiermb4 --ip x --token y info
Model: zhimi.airpurifier.mb4
Hardware version: esp32
Firmware version: 2.0.8
 ~/python-miio/miio   add-mb4-support ± python3 cli.py airpurifiermb4 --ip x --token y status
Traceback (most recent call last):
  File "/home/jagub2/python-miio/miio/cli.py", line 51, in <module>
    create_cli()
  File "/home/jagub2/python-miio/miio/cli.py", line 47, in create_cli
    return cli(auto_envvar_prefix="MIIO")
  File "/home/jagub2/python-miio/miio/../miio/click_common.py", line 59, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/jagub2/python-miio/miio/../miio/click_common.py", line 285, in wrap
    result_msg = result_msg_fmt.format(**kwargs)
AttributeError: 'AirPurifierMB4Status' object has no attribute 'favorite_rpm'
 ✘  ~/python-miio/miio   add-mb4-support ± python3 cli.py airpurifiermb4 --ip x --token y set_favorite_rpm 300
Setting favorite motor speed '300' rpm
[{'did': 'favorite_rpm', 'siid': 9, 'piid': 3, 'code': 0}]
 ~/python-miio/miio   add-mb4-support ± python3 cli.py airpurifiermb4 --ip x --token y status
Traceback (most recent call last):
  File "/home/jagub2/python-miio/miio/cli.py", line 51, in <module>
    create_cli()
  File "/home/jagub2/python-miio/miio/cli.py", line 47, in create_cli
    return cli(auto_envvar_prefix="MIIO")
  File "/home/jagub2/python-miio/miio/../miio/click_common.py", line 59, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/jagub2/python-miio/miio/../miio/click_common.py", line 285, in wrap
    result_msg = result_msg_fmt.format(**kwargs)
AttributeError: 'AirPurifierMB4Status' object has no attribute 'favorite_rpm'
```

set_favorite_rpm makes it beep but it still fails on `status` command. All other set_ commands work, too besides set_led_brightness_level, for that one I get:

```
Traceback (most recent call last):
  File "/home/jagub2/python-miio/miio/cli.py", line 49, in <module>
    create_cli()
  File "/home/jagub2/python-miio/miio/cli.py", line 45, in create_cli
    return cli(auto_envvar_prefix="MIIO")
  File "/home/jagub2/python-miio/miio/click_common.py", line 59, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/jagub2/python-miio/miio/click_common.py", line 277, in wrap
    msg = msg_fmt.format(**kwargs)
KeyError: 'brightness'
```

When I hardcode `favorite_rpm` in `default_output` in `AirPurifierMB4`'s `__init__`, `status` command starts working:
```
Power: on
AQI: 4 μg/m³
Mode: OperationMode.Favorite
LED brightness level: 3
Buzzer: False
Child lock: True
Filter life remaining: 99 %
Filter hours used: 29
Motor speed: 300 rpm
Favorite RPM: 300 rpm
```

I checked out that project and exported `PYTHONPATH` to use this module without installing.

Am I doing something wrong? Am I missing something? Thanks in advance.

**Comment by arturdobo** [comment]
@jagub2 I'm going to apply the suggestions above tomorrow, so I'll check what's wrong then. I tested it using HA but maybe omitted something on the CLI 🤔  Thanks for letting me know

**Comment by arturdobo** [comment]
@jagub2 could you pull changes, install and check everything is ok now?

File: miio/airpurifier_miot.py
Comment (arturdobo) [reply miio/airpurifier_miot.py]: done

File: miio/airpurifier_miot.py
Comment (arturdobo) [reply miio/airpurifier_miot.py]: done

**Comment by jagub2** [comment]
@arturdobo I confirm, everything works great now. Thank you!

**Comment by mbatorowicz** [comment]
Hello, I also have the 3C version. Can you please explain step by step what I have to do to make it work with HA.

File: miio/airpurifier_miot.py
Comment (rytilahti) [inline miio/airpurifier_miot.py:364]:
```suggestion
```
I don't think this is necessary here, as it does nothing calling the parent class `__init__` would do?

File: miio/airpurifier_miot.py
Comment (rytilahti) [inline miio/airpurifier_miot.py:567]:
```suggestion
        """Set led brightness level (0..8)."""
```

File: miio/airpurifier_miot.py
Comment (rytilahti) [inline miio/airpurifier_miot.py:149]:
```suggestion
```

**Review by rytilahti — CHANGES_REQUESTED** [review]
Just a couple of minor issues to be fixed, then this is ready to go! :+1: Please also update the supported devices listing in README!

**Comment by arturdobo** [comment]
@rytilahti all suggestions have been applied. Regarding `README` how to add information about the new air purifier? As a new entry in the list or somehow different? I see there is a generic entry `Xiaomi Mi Air Purifier`, but no idea how many models it covers

**Comment by rytilahti** [comment]
@arturdobo Maybe just adding `(3C, ..)` to the end of the entry, I think we want to have a yaml-based registry of supported devices & their model information at some point, but until then it's manual updates to the readme :(

**Comment by arturdobo** [comment]
@mbatorowicz I'm definitively not a python developer so I did it a hacky way at the beginning of development by replacing the class in the `python-miio` lib. But once this PR is merged you should be able to just update the library version and test.

@rytilahti I added a whole list of air purifiers I use and I'm sure `python-miio` supports them.

**Comment by rytilahti** [comment]
LGTM, let's get it merged. Thanks again! :tada: 

> @rytilahti I added a whole list of air purifiers I use and I'm sure `python-miio` supports them.

Great! Oh, you have multiple purifiers? Could you create a separate PR to add a list of model identifiers to the docstrings of the respective implementation classes?

**Review by rytilahti — APPROVED** [review]

**Comment by arturdobo** [comment]
Yeah! One air purifier per room, but I wanted to replace the old `2` model with `3C` model so temporarily I have one more than needed 😄 

Sure, I'll add those models to appropriate docstrings

**Comment by gh0st125** [comment]
I am still not able to integrate my Air Purifier 3C into Home Assistant. The Entity always states "not available"

image

This is what i have in my configuration.yaml:

`fan:
  - platform: xiaomi_miio
    host: 192.168.0.53
    token: 96cc1e45d258fc291998b9c0848877e2
    name: "Luftreiniger"`

what am I missing?

**Comment by arturdobo** [comment]
@gh0st125 is it your only change in HA? If yes, it won't work like that. More details [here](https://github.com/syssi/xiaomi_airpurifier/issues/117#issuecomment-776162370)

**Comment by gh0st125** [comment]
yeah, these are the only changes so far. 

> Feel free to provide a pull request for this custom component as first step to support your device. I'm trying to refactor the component to be merged as official HA component again some day.

So I have to do a pull request? How do I do that? Sorry, I'm not really familiar with this kind of integrations...

**Comment by arturdobo** [comment]
Once @rytilahti releases a new version of `python-miio` the device can be added to HA. The way new Xiaomi devices were added so far is not allowed anymore and the new devices need to be added the new way meaning the one device needs to be split into multiple entities, one per measurement (i.e. humidity, air quality, etc). Any PR that adds devices (or even features to already added devices) the old way are immediately closed in HA 😉

**Comment by gh0st125** [comment]
ahh okay thanks, so I have to wait for a new release. That's fine for me, it's not something that I have to have right now :D Thanks for the clarification!

**Comment by arturdobo** [comment]
I'll try to prepare some PR in HA and see how it will work 😉

**Comment by calas80** [comment]
I have to wait for a new release. Thanks

**Comment by starkillerOG** [comment]
@arturdobo I am currently working on implementing Config Flow for all Miio platforms.
Switches are done, Vacuums almost done (only needs to be merged).
Lights are up next, but I could also do the fan platform (Air Purifier and Humidifier) first if you like.

Probably you will want to wait untill the Config Flow is done before implementing this new device, but of course it is up to you.

**Comment by arturdobo** [comment]
it'd be great if you could.

**Comment by starkillerOG** [comment]
@arturdobo just made the PR: https://github.com/home-assistant/core/pull/46866

**Comment by abrilevskiy** [comment]
Could you please clarify - can zhimi.airpurifier.mb4 be added to HomeAssistant? I tried with the latest Dev without success

**Comment by starkillerOG** [comment]
@abrilevskiy the config flow PR is merged, but the zhimi.airpurifier.mb4 has not been implemented.
However now that the config flow PR is merged, someone could proceed with making a PR for the zhimi.airpurifier.mb4.

**Comment by arturdobo** [comment]
@rytilahti could you please release the new `python-miio` version? It allows adding new devices do HA then

**Comment by jddipqd** [comment]
Is 3C supported now?
I'm running 0.5.5.2 but still can't get it to work.
```
miiocli airpurifier --ip xxx --token xxx info

Model: zhimi.airpurifier.mb4
Hardware version: esp32
Firmware version: 2.0.8
```

```
miiocli airpurifier --ip xxx --token xxx status

Error: {'code': -5000, 'message': 'method not found'}
```

The same command is working for 2H.

**Comment by rytilahti** [comment]
I think it should be supported as per this PR? Feel free to open a separate issue if that's not the case.

**Comment by arturdobo** [comment]
The argument you provide is wrong, `airpurifiermb4` should be instead

**Comment by jddipqd** [comment]
Thank you, it's working with `airpurifiermb4`.

**Pull request merged**
